version: "3.9"

services:
  db:
    image: postgres:14.1
    container_name: postgres
    ports:
      - 5432:5432
    environment:
      - POSTGRES_DB=social_media
      - POSTGRES_USER=milad73n
      - POSTGRES_PASSWORD=14863654

  rabbitmq:
    image: "rabbitmq:alpine"
    container_name: rabbitmq
    healthcheck:
      test: rabbitmq-diagnostics -q ping
      interval: 30s
      timeout: 30s
      retries: 3

  django:
    build:
      context: .
      dockerfile: docker/production.Dockerfile
    container_name: django
    command: ./docker/web_entrypoint.sh
    environment:
      - DATABASE_URL=psql://milad73n:14863654@db:5432/social_media
      - CELERY_BROKER_URL=amqp://guest:guest@rabbitmq:5672//
    volumes:
      - .:/app
    ports:
      - "8000:8000"
    depends_on:
      - db
      - rabbitmq
    restart: on-failure

  celery:
    build:
      context: .
      dockerfile: docker/production.Dockerfile
    # command: celery -A social_media.tasks worker -l info --without-gossip --without-mingle --without-heartbeat
    container_name: worker
    command: ./docker/celery_entrypoint.sh
    environment:
      - DATABASE_URL=psql://milad73n:14863654@db:5432/social_media
      - CELERY_BROKER_URL=amqp://guest:guest@rabbitmq:5672//
    volumes:
      - .:/app
    depends_on:
      - db
      - rabbitmq
    restart: on-failure

  beats:
    build:
      context: .
      dockerfile: docker/production.Dockerfile
    # command: celery -A social_media.tasks beat -l info --scheduler django_celery_beat.schedulers:DatabaseScheduler
    container_name: beats
    command: ./docker/beats_entrypoint.sh
    environment:
      - DATABASE_URL=psql://milad73n:14863654@db:5432/social_media
      - CELERY_BROKER_URL=amqp://guest:guest@rabbitmq:5672//
    volumes:
      - .:/app
    depends_on:
      - db
      - rabbitmq
    restart: on-failure
